File: cpp/template/topo_sort.cpp
// Template 1. topo sort using BFS
// O(N) / O(N)
class Solution {
public:
    int canFinish(int n, vector<vector<int>>& prereq) {
        // constuct graph
        vector<vector<int>> graph(n);
        vector<int> in_degree(n, 0), sources;
        for (auto& e : prereq){
            graph[e[1]].push_back(e[0]);
            in_degree[e[0]]++;
        }
        
        // find all sources
        for (int i = 0; i < n; ++i) 
            if (!in_degree[i]) sources.push_back(i);
        
        // bfs
        int cnt = 0;
        while (!sources.empty()) {
            vector<int> nxt_sources;
            for (int cur : sources) {
                cnt++;
                for (int nxt : graph[cur]) {
                    if (--in_degree[nxt] == 0)
                        nxt_sources.push_back(nxt);
                }
            }
            sources = nxt_sources;
        }
        
        return cnt;
    }
};


// Template 2. topo sort using DFS
// O(N) / O(N)
class Solution {
public:
    int canFinish(int n, vector<vector<int>>& prereq) {
        // constuct graph
        vector<vector<int>> graph(n);
        vector<int> in_degree(n, 0), sources;
        for (auto& e : prereq){
            graph[e[1]].push_back(e[0]);
            in_degree[e[0]]++;
        }
        
        // find all sources
        for (int i = 0; i < n; ++i) 
            if (!in_degree[i]) sources.push_back(i);
        
        // dfs
        int cnt = 0;
        while (!sources.empty()) {
            cnt++;
            int cur = sources.back();
            sources.pop_back();
            for (int nxt : graph[cur]) {
                if (--in_degree[nxt] == 0)
                    sources.push_back(nxt);
            }
        }
        
        return cnt;
    }
};

File: python/template/topo_sort.py
# Template 1. topo sort using BFS
# O(N) / O(N)
class Solution:
    def canFinish(self, n, prereq):
        # construct graph
        graph = {i: [] for i in range(n)}
        in_degree = {i: 0 for i in range(n)}
        for u, v in prereq:
            graph[v].append(u)
            in_degree[u] += 1
        
        # find all sources
        sources = [i for i in range(n) if in_degree[i] == 0]
        
        # bfs
        cnt = 0
        while sources:
            nxt_sources = []
            for cur in sources:
                cnt += 1
                for nxt in graph[cur]:
                    in_degree[nxt] -= 1
                    if in_degree[nxt] == 0:
                        nxt_sources.append(nxt)
            sources = nxt_sources
            
        return cnt
        

# Template 2. topo sort using DFS
# O(N) / O(N)
class Solution:
    def canFinish(self, n, prereq):
        # construct graph
        graph = {i: [] for i in range(n)}
        in_degree = {i: 0 for i in range(n)}
        for u, v in prereq:
            graph[v].append(u)
            in_degree[u] += 1
        
        # find all sources
        sources = [i for i in range(n) if in_degree[i] == 0]
        
        # dfs
        cnt = 0
        while sources:
            cur = sources.pop()
            cnt += 1
            for nxt in graph[cur]:
                in_degree[nxt] -= 1
                if in_degree[nxt] == 0:
                    sources.append(nxt)
                
        return cnt